Source/WebCore/platform/graphics/PixelBuffer.cpp

```diff
@@ -38,30 +38,58 @@ bool PixelBuffer::supportedPixelFormat(PixelFormat pixelFormat)
     switch (pixelFormat) {
     case PixelFormat::RGBA8:
     case PixelFormat::BGRA8:
+#if HAVE(HDR_SUPPORT)
+    case PixelFormat::RGBA16F:
+#endif
         return true;
 
     case PixelFormat::BGRX8:
     case PixelFormat::RGB10:
     case PixelFormat::RGB10A8:
-#if HAVE(HDR_SUPPORT)
-    case PixelFormat::RGBA16F:
-#endif
         return false;
     }
 
     ASSERT_NOT_REACHED();
     return false;
 }
 
-CheckedUint32 PixelBuffer::computeBufferSize(PixelFormat pixelFormat, const IntSize& size)
+static CheckedUint32 mustFitInInt32(CheckedUint32 uint32)
+{
+    if (!uint32.hasOverflowed() && uint32.value() > uint32_t(std::numeric_limits<int32_t>::max()))
+        uint32.overflowed();
+    return uint32;
+}
+
+static CheckedUint32 computeRawPixelCount(const IntSize& size)
+{
+    return CheckedUint32 { size.width() } * size.height();
+}
+
+static CheckedUint32 computeRawPixelComponentCount(PixelFormat pixelFormat, const IntSize& size)
+{
+    ASSERT_UNUSED(pixelFormat, PixelBuffer::supportedPixelFormat(pixelFormat));
+    constexpr unsigned componentsPerPixel = 4;
+    return computeRawPixelCount(size) * componentsPerPixel;
+}
+
+CheckedUint32 PixelBuffer::computePixelCount(const IntSize& size)
+{
+    return mustFitInInt32(computeRawPixelCount(size));
+}
+
+CheckedUint32 PixelBuffer::computePixelComponentCount(PixelFormat pixelFormat, const IntSize& size)
 {
-    ASSERT_UNUSED(pixelFormat, supportedPixelFormat(pixelFormat));
-    constexpr unsigned bytesPerPixel = 4;
-    auto bufferSize = CheckedUint32 { size.width() } * size.height() * bytesPerPixel;
-    if (!bufferSize.hasOverflowed() && bufferSize.value() > std::numeric_limits<int32_t>::max())
-        bufferSize.overflowed();
-    return bufferSize;
+    return mustFitInInt32(computeRawPixelComponentCount(pixelFormat, size));
+}
 
+CheckedUint32 PixelBuffer::computeBufferSize(PixelFormat pixelFormat, const IntSize& size)
+{
+    unsigned bytesPerPixelComponent =
+#if HAVE(HDR_SUPPORT)
+        (pixelFormat == PixelFormat::RGBA16F) ? 2 :
+#endif
+        1;
```

Comment:
I thought there might be an existing function around that does this, but I couldn't find one.

Reply:
Yeah, same. Ideally we'd want some table/function that knows the sizeof of each pixel format, but I decided not to, because:
- We'd need to work out the sizeof of all possible formats, even if it probably wouldn't be used anywhere(?).
- I'm not sure what it would be for (unsupported) RGB10 and RGB10A8. 😅 Or does it even make sense, if components are packed in bit ranges inside a bigger pixel bundle? Anyway, too much to worry about when it's not needed here.

I'll add a FIXME to revisit.